src/control/server/storage_nvme.go

```diff
@@ -183,53 +208,128 @@ func (n *nvmeStorage) Discover() error {
+//
+// One result with empty Pciaddr will be reported if there are preliminary
+// errors occurring before devices could be accessed. Otherwise a result will
+// be populated for each device in bdev_list.
+func (n *nvmeStorage) Format(i int, resp *pb.FormatStorageResp) {
+	srv := n.config.Servers[i]
+
+	// firstResult reports any preliminary errors
+	firstResult := &pb.NvmeControllerResult{
+		Pciaddr: "",
+		State:   new(pb.ResponseState),
+	}
+	firstState := firstResult.State
+
 	if !n.initialized {
-		return errors.New("nvme storage not initialized")
+		firstState.Status = pb.ResponseStatus_CTRL_ERR_APP
+		firstState.Error = ErrNvmeNotInited
+		resp.Crets = append(resp.Crets, firstResult)
+		return
 	}
+
 	if n.formatted {
-		return errors.New(
-			"nvme storage has already been formatted and reformat " +
-				"not implemented")
+		firstState.Status = pb.ResponseStatus_CTRL_ERR_APP
+		firstState.Error = ErrNvmeFormatted
+		resp.Crets = append(resp.Crets, firstResult)
+		return
 	}
 
-	srv := n.config.Servers[idx]
-
 	switch srv.BdevClass {
 	case bdNVMe:
 		for _, pciAddr := range srv.BdevList {
+			// generate one result per list entry
+			resp.Crets = append(
+				resp.Crets,
+				&pb.NvmeControllerResult{
+					Pciaddr: pciAddr,
+					State:   new(pb.ResponseState),
+				})
+
+			// ref to state of last result addetd
+			state := resp.Crets[len(resp.Crets)-1].State
+
 			if pciAddr == "" {
-				return errors.New("bdev nvme device list entry empty")
+				state.Status = pb.ResponseStatus_CTRL_ERR_CONF
+				state.Error = ErrBdevEmpty
+				continue
+			}
+			resp.Crets[len(resp.Crets)-1].Pciaddr = pciAddr
+
+			ctrlr := n.getController(pciAddr)
+			if ctrlr == nil {
+				state.Status = pb.ResponseStatus_CTRL_ERR_NVME
+				state.Error = pciAddr + ": " + ErrBdevNotFound
+				continue
 			}
 
 			cs, ns, err := n.nvme.Format(pciAddr)
 			if err != nil {
-				return errors.Wrap(err, "nvme format")
+				state.Status = pb.ResponseStatus_CTRL_ERR_NVME
+				state.Error = pciAddr + ": " + err.Error()
+				continue
 			}
 			n.controllers = loadControllers(cs, ns)
 		}
 	default:
-		return errors.Errorf(
-			"format unsupported on BdevClass %v", srv.BdevClass)
+		firstState.Status = pb.ResponseStatus_CTRL_ERR_CONF
+		firstState.Error = string(srv.BdevClass) + ": " +
+			ErrBdevClassNotSupported
+		resp.Crets = append(resp.Crets, firstResult)
+		return
 	}
 
 	n.formatted = true
-	return nil
+	return
 }
 
-// Update method implementation for nvmeStorage
-func (n *nvmeStorage) Update(pciAddr string, path string, slot int32) error {
+// Update method implementation for nvmeStorage attempts to update firmware on
+// NVMe controllers attached to a given server identified by PCI addresses as
+// specified in per-server section of the config file.
+//
+// Firmware will only be updated if the controller reports the current fw rev
+// to be the same as the "startRev" fn parameter. path and slot refer to the
```

Comment:
in discussion with customer decided upon only performing update on known upgrade paths, i.e. from v1 to v2. so upgrade will only occur if the controller is already at v1 (startRev). The target version will be that of the image provided. that any more clear? I will try to improve the wording.

Reply:
That use case makes sense to me - I just find the enforced limitation a little surprising. I guess I just don't totally understand the use case where the user is running with mixed FW versions, and doesn't want to take all drives to the same version on the next update.

Reply:
I think it's because there is some concern that some upgrade paths are not supported (and may cause damage),therefore we want to only follow known good upgrade paths

Reply:
Sure, that makes sense.